enhancements/network/bgp-overview.md

```diff
@@ -0,0 +1,172 @@
+---
+title: bgp-overview
+authors:
+  - "@russellb"
+reviewers:
+  - "@danwinship"
+  - "@squeed"
+approvers:
+  - TBD
+creation-date: 2020-08-11
+last-updated: 2020-08-11
+status: informational
+see-also:
+  - https://github.com/openshift/enhancements/pull/356
+  - https://github.com/openshift/enhancements/pull/394
+---
+
+# OpenShift and BGP
+
+BGP (Border Gateway Protocol) is a dynamic routing protocol used to power a
+number of important networking use cases, including routing in the global
+internet.  For some helpful resources on learning more about BGP, see the
+[BGP Resources](#bgp-resources) section of this document.
+
+The purpose of this document is to serve as a high level overview of how BGP
+relates to OpenShift.  We will review a number of use cases, discuss how they
+could be supported, and link to existing related work.  Any future enhancements
+related to BGP should be able to rely on this one to provide some broader
+context and should continue to update this document as necessary.
+
+## Use Cases
+
+### L3 (Layer 3) Network Fabric
+
+The use case here does not require any changes to OpenShift.  It is still
+useful to call out as it is one use of BGP that is relevant to OpenShift,
+particularly with on premise clusters.
+
+It is possible to use BGP between the routers used to build the L3 network that
+OpenShift Nodes are connected to.
+
+### Dynamic Routing for Egress Traffic
+
+*First, an aside on terminology: This is referring to IP Routing, and
+not the OpenShift feature called Routing that works at the HTTP layer.  There
+are also OpenShift features that use the word "egress", but in this context, we
+only mean to refer to traffic that is leaving the cluster and not related to
+any other features that use that name.*
+
+Some networks may have more than one router accessible from Nodes for
+availability reasons.  In this case an administrator may want to use BGP down
+to the Nodes themselves so they are aware of the multiple routing options and
+potentially use ECMP (Equal Cost Multi Path) to distribute traffic
+across each of the routers.
+
+This capability is not provided by OpenShift today.  It may be possible as a
+custom configuration, but not with all network providers.  With OpenShift-SDN,
+it should be possible to run your own BGP routing daemon on the host via a
+`DaemonSet`.
+
+OVN-Kubernetes uses a different host network configuration where egress
+traffic is routed by Open vSwitch flows programmed by OVN.  It means that a
+dynamic routing daemon must also feed those routes into OVN configuration, so
+it's not as straight forward as just running an existing daemon on the host.
+
+A future enhancement is needed to describe how we would provide this
+functionality natively with OpenShift.
+
+### Pod Network Control Plane
```

Comment:
So there are two parts to what Calico does with BGP:

1. knowing that Pod A is on Node X
2. actually routing traffic for Pod A to Node X

You're only talking about 1 here, but I think 2 is the more interesting / more discussed aspect. OpenShift SDN and OVN Kubernetes implement 1 differently, but they both implement 2 the same way, by encapsulating inter-node pod traffic in a tunnel. In at lease some configurations of Calico, they implement 2 by ensuring that each node can simply route traffic to pods on other nodes directly, thereby avoiding the need for encapsulation.

Reply:
Yes, great points.  How about I add a new section that discusses "avoiding tunnel encapsulation" ?

Reply:
sure

Reply:
+1 - I was trying to figure out whether "BGP for pod network control plane" means "no encapsulation", and it didn't sound like it. I'm glad I wasn't missing something obvious!

Reply:
I've pushed another commit which includes a new section about avoiding tunnel encapsulation.